internal/rpc/clients.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpc
+
+import (
+	"crypto/tls"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"open-match.dev/open-match/internal/config"
+)
+
+var (
+	clientLogger = logrus.WithFields(logrus.Fields{
+		"app":       "openmatch",
+		"component": "client",
+	})
+)
+
+// ClientParams contains the connection parameters to connect to an Open Match service.
+type ClientParams struct {
+	Hostname          string
+	Port              int
+	PublicCertificate []byte
+	PrivateKeyData    []byte
+}
+
+func (p *ClientParams) usingTLS() bool {
+	return len(p.PublicCertificate) > 0
+}
+
+// GRPCClientFromConfig creates a gRPC client connection from a configuration.
+func GRPCClientFromConfig(cfg config.View, prefix string) (*grpc.ClientConn, error) {
+	hostname := cfg.GetString(prefix + ".hostname")
+	portNumber := cfg.GetInt(prefix + ".grpcport")
+
+	// If TLS support is not enabled in the config, return an insecure grpc client.
+	if !cfg.GetBool("tls.enabled") {
+		return GRPCClientFromParams(&ClientParams{
+			Hostname: hostname,
+			Port:     portNumber,
+		})
+	}
+
+	// If TLS support is enabled, read and return a grpc TLS client.
+	publicKeyPath := cfg.GetString("tls.publicKeyPath")
+
+	publicKeyBytes, err := ioutil.ReadFile(publicKeyPath)
+	if err != nil {
+		clientLogger.WithError(err).Fatalf("cannot read public key from file %s", publicKeyPath)
+		return nil, err
+	}
+
+	return GRPCClientFromParams(&ClientParams{
+		Hostname:          hostname,
+		Port:              portNumber,
+		PublicCertificate: publicKeyBytes,
```

Comment:
Seems weird to me that public key would be required here.  What public key is this?

Reply:
> Seems weird to me that public key would be required here. What public key is this?

This is actually the public key from the one who signed the certificate. Client will need to use it to decrypt the signature of server certificate. I'll probably rename it to PublicKeyForCertificate